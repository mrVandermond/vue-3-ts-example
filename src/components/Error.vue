<template>
  <div
    :class="['error', {
      'error__title_right': props.isTitleFromRight,
    }]"
  >
    <img
      src="assets/error.png"
      :class="['error-image', {
        'error-image_without-margin': props.isTitleFromRight,
      }]"
      alt="error-image"
    >

    <div
      :class="{
        'title_position_right': props.isTitleFromRight,
      }"
    >
      <p class="title-text title-text__bold">
        Сервер не отвечает
      </p>

      <p class="title-text">
        Уже работаем над этим
      </p>
    </div>
  </div>
</template>

<script lang="ts" setup>
const props = withDefaults(defineProps<{
  isTitleFromRight?: boolean;
}>(), {
  isTitleFromRight: false,
});
</script>

<style lang="scss">
.error-image {
  margin-bottom: 32px;

  &_without-margin {
    margin-bottom: 0;
  }
}
.error {
  display: flex;
  align-items: center;
  justify-content: center;
  flex-wrap: wrap;
  max-width: 264px;
}
.error__title_right {
  max-width: initial;
}
.title_position_right {
  margin-left: 24px;
}
</style>
