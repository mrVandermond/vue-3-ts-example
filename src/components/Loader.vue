<template>
  <img
    class="loader"
    src="assets/loader.gif"
    alt="loader-image"
  >
</template>

<script lang="ts" setup>
</script>

<style lang="scss">
.loader {
  width: 48px;
  height: 48px;
}
</style>
