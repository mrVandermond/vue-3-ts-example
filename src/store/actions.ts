import { storeTypes } from '@/store/types';
import { utlCst } from '@/utils';
import type { Mutations } from '@/store/mutations';
import type { ActionContext, ActionTree } from 'vuex';
import type { IState } from '@/store/state';
import type { IAlbum, IPhoto, IPhotoWithIsFavorite, IUser } from '@/types';
import * as api from '@/api';

type AugmentedActionContext = {
  commit<K extends keyof Mutations>(
    key: K,
    payload: Parameters<Mutations[K]>[1]
  ): ReturnType<Mutations[K]>;
} & Omit<ActionContext<IState, IState>, 'commit'>;

export interface Actions {
  [storeTypes.EActions.GET_FAVORITE_PHOTOS_FROM_LOCAL_STORAGE](
    ctx: AugmentedActionContext,
  ): void;
  [storeTypes.EActions.SET_FAVORITE_PHOTOS_TO_LOCAL_STORAGE](
    ctx: AugmentedActionContext,
  ): void;
  [storeTypes.EActions.FETCH_USERS](
    ctx: AugmentedActionContext,
  ): Promise<IUser[]>;
  [storeTypes.EActions.FETCH_ALBUMS_BY_USER_ID](
    ctx: AugmentedActionContext,
    payload: number,
  ): Promise<IAlbum[]>;
  [storeTypes.EActions.FETCH_PHOTOS_BY_ALBUM_ID](
    ctx: AugmentedActionContext,
    payload: number,
  ): Promise<IPhoto[]>;
}

export const actions: ActionTree<IState, IState> & Actions = {
  async [storeTypes.EActions.FETCH_USERS]() {
    const { data } = await api.fetchUsers();

    return data;
  },
  async [storeTypes.EActions.FETCH_ALBUMS_BY_USER_ID](ctx, userId) {
    const { data } = await api.fetchAlbumsByUserId(userId);

    return data;
  },
  async [storeTypes.EActions.FETCH_PHOTOS_BY_ALBUM_ID](ctx, albumId) {
    const { data } = await api.fetchPhotosByAlbumId(albumId);

    return data;
  },
  [storeTypes.EActions.GET_FAVORITE_PHOTOS_FROM_LOCAL_STORAGE]({ commit }) {
    const jsonFavoritePhotos = localStorage.getItem(utlCst.LOCAL_STORAGE_KEY);

    if (!jsonFavoritePhotos) return;

    const favoritePhotos = (JSON.parse(jsonFavoritePhotos) as IPhotoWithIsFavorite[]);

    commit(storeTypes.EMutations.ADD_TO_FAVORITES, favoritePhotos);
  },
  [storeTypes.EActions.SET_FAVORITE_PHOTOS_TO_LOCAL_STORAGE]({ state }) {
    const { favoritePhotos } = state;
    const jsonFavoritePhotos = JSON.stringify(favoritePhotos);

    localStorage.setItem(utlCst.LOCAL_STORAGE_KEY, jsonFavoritePhotos);
  },
};
