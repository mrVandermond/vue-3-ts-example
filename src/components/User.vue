<template>
  <ListItem
    :is-loading-content="isLoadingAlbums"
    :is-fail-loading-content="isFailLoadingAlbums"
    @click="onClickItem"
  >
    <template #label>
      <div class="user-label">
        {{ props.user.name }}
      </div>
    </template>

    <template
      v-if="isOpenUser"
      #content
    >
      <Album
        v-for="album in albums"
        :key="album.id"
        :album="album"
      />
    </template>
  </ListItem>
</template>

<script lang="ts" setup>
import ListItem from '@/components/ListItem.vue';
import Album from '@/components/Album.vue';
import type { IAlbum, IUser } from '@/types';
import { ref } from 'vue';
import { customUseStore } from '@/store';
import { storeTypes } from '@/store/types';

const props = defineProps<{
  user: IUser;
}>();
const store = customUseStore();
const albums = ref<IAlbum[]>([]);
const isOpenUser = ref(false);
const isLoadingAlbums = ref(false);
const isFailLoadingAlbums = ref(false);

const onClickItem = async (isOpen: boolean): Promise<void> => {
  isOpenUser.value = isOpen;

  if (!isOpen) return;

  isLoadingAlbums.value = true;

  try {
    albums.value = await store.dispatch(storeTypes.EActions.FETCH_ALBUMS_BY_USER_ID, props.user.id);
  } catch (error) {
    isFailLoadingAlbums.value = true;
  }

  isLoadingAlbums.value = false;
};
</script>

<style lang="scss">
.user-label {
  font-weight: 500;
  font-size: 22px;
  height: 22px;
}
</style>
