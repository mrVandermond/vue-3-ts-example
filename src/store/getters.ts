import type { GetterTree } from 'vuex';
import type { IState } from '@/store/state';
import { storeTypes } from '@/store/types';

export interface Getters {
  [storeTypes.EGetters.IS_FAVORITE_PHOTO_BY_ID](state: IState): (photoId: number) => boolean;
}

export const getters: GetterTree<IState, IState> & Getters = {
  [storeTypes.EGetters.IS_FAVORITE_PHOTO_BY_ID]: (state) => (photoId: number) => {
    const photo = state.favoritePhotos.find((item) => item.id === photoId);

    return !!photo;
  },
};
