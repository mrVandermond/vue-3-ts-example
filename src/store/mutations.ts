import { storeTypes } from '@/store/types';
import type { IModalConfig, IPhotoWithIsFavorite, ITitleConfig } from '@/types';
import type { IState } from '@/store/state';
import type { MutationTree } from 'vuex';
import { utlFuncs } from '@/utils';

export interface Mutations<S = IState> {
  [storeTypes.EMutations.ADD_TO_FAVORITES](state: S, payload: IPhotoWithIsFavorite[]): void;
  [storeTypes.EMutations.REMOVE_FROM_FAVORITES](state: S, payload: IPhotoWithIsFavorite): void;
  [storeTypes.EMutations.SET_MODAL_CONFIG](state: S, payload: IModalConfig): void;
  [storeTypes.EMutations.SET_TITLE_CONFIG](state: S, payload: ITitleConfig): void;
}

export const mutations: MutationTree<IState> & Mutations = {
  [storeTypes.EMutations.ADD_TO_FAVORITES](state, photos: IPhotoWithIsFavorite[]) {
    const photosWithoutExisting = photos.filter((photo) => (
      utlFuncs.getItemIndexBy(state.favoritePhotos, photo, 'id') === -1
    ));

    state.favoritePhotos.push(...photosWithoutExisting.map((photo) => ({ ...photo, isFavorite: true })));
  },
  [storeTypes.EMutations.REMOVE_FROM_FAVORITES](state, photo: IPhotoWithIsFavorite) {
    const photoIndex = utlFuncs.getItemIndexBy(state.favoritePhotos, photo, 'id');

    if (photoIndex === -1) return;

    state.favoritePhotos.splice(photoIndex, 1);
  },
  [storeTypes.EMutations.SET_MODAL_CONFIG](state, config) {
    state.modal.isOpen = config.isOpen;
    state.modal.url = config.url;
  },
  [storeTypes.EMutations.SET_TITLE_CONFIG](state, config) {
    state.title.isShow = config.isShow;
    state.title.text = config.text;
    state.title.xCoord = config.xCoord;
    state.title.yCoord = config.yCoord;
  },
};
