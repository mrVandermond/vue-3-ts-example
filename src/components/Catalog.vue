<template>
  <div>
    <Loader v-if="isLoading" />

    <Error v-else-if="!isLoading && isFailLoading" />

    <template v-else>
      <User
        v-for="user in users"
        :key="user.id"
        :user="user"
        class="user"
      />
    </template>
  </div>
</template>

<script lang="ts" setup>
import User from '@/components/User.vue';
import Loader from '@/components/Loader.vue';
import Error from '@/components/Error.vue';
import { onBeforeMount, onBeforeUnmount, ref } from 'vue';
import type { IUser } from '@/types';
import * as types from '@/store/types';
import { customUseStore } from '@/store';
import { storeTypes } from '@/store/types';

const isLoading = ref(false);
const isFailLoading = ref(false);
const users = ref<IUser[]>([]);
const store = customUseStore();

onBeforeMount(async (): Promise<void> => {
  await store.dispatch(types.EActions.GET_FAVORITE_PHOTOS_FROM_LOCAL_STORAGE);

  isLoading.value = true;

  try {
    users.value = await store.dispatch(storeTypes.EActions.FETCH_USERS);
  } catch (error) {
    isFailLoading.value = true;
  }

  isLoading.value = false;
});
onBeforeUnmount(async (): Promise<void> => {
  await store.dispatch(types.EActions.SET_FAVORITE_PHOTOS_TO_LOCAL_STORAGE);
});

</script>

<style lang="scss">
.user {
  width: 100%;
}
</style>
