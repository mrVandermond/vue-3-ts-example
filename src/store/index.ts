import type { StoreOptions } from 'vuex';
import { createStore, useStore as baseUseStore } from 'vuex';
import type { InjectionKey } from 'vue';
import type { Store } from '@/store/types/overrides';
import type { IState } from './state';
import { state } from './state';
import { mutations } from './mutations';
import { getters } from './getters';
import { actions } from './actions';

export const key: InjectionKey<Store> = Symbol('store');

export function customUseStore(): Store {
  return baseUseStore(key);
}

export function customCreateStore(options: StoreOptions<IState>): Store {
  return createStore<IState>(options);
}

export const store = customCreateStore({
  state,
  mutations,
  getters,
  actions,
});
