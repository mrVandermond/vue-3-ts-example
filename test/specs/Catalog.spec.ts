import Catalog from '@/components/Catalog.vue';
import Loader from '@/components/Loader.vue';
import Error from '@/components/Error.vue';
import User from '@/components/User.vue';
import { key, customCreateStore } from '@/store';
import { storeTypes } from '@/store/types';

import { nextTick } from 'vue';
import type { DefineComponent } from 'vue';
import type { VueWrapper } from '@vue/test-utils';
import type { Store } from '@/store/types/overrides';
import type { IUser } from '@/types';

import { flushPromises } from '@vue/test-utils';
import { state } from '@/store/state';
import { mutations } from '@/store/mutations';
import { getters } from '@/store/getters';
import { actions } from '@/store/actions';
import { wrapperFactory } from '../utils';

describe('Catalog component', () => {
  let store!: Store;
  const exampleUsers: IUser[] = [
    {
      id: 1,
      name: 'Leanne Graham',
      username: 'Bret',
      email: 'Sincere@april.biz',
      address: {
        street: 'Kulas Light',
        suite: 'Apt. 556',
        city: 'Gwenborough',
        zipcode: '92998-3874',
      },
      phone: '1-770-736-8031 x56442',
      website: 'hildegard.org',
      company: {
        name: 'Romaguera-Crona',
        catchPhrase: 'Multi-layered client-server neural-net',
        bs: 'harness real-time e-markets',
      },
    },
    {
      id: 2,
      name: 'Ervin Howell',
      username: 'Antonette',
      email: 'Shanna@melissa.tv',
      address: {
        street: 'Victor Plains',
        suite: 'Suite 879',
        city: 'Wisokyburgh',
        zipcode: '90566-7771',
      },
      phone: '010-692-6593 x09125',
      website: 'anastasia.net',
      company: {
        name: 'Deckow-Crist',
        catchPhrase: 'Proactive didactic contingency',
        bs: 'synergize scalable supply-chains',
      },
    },
  ];

  beforeEach(() => {
    store = customCreateStore({
      state,
      mutations,
      getters,
      actions,
    });
  });

  const getWrapper = (): VueWrapper<InstanceType<DefineComponent>> => wrapperFactory(Catalog, {
    global: {
      plugins: [[store, key]],
      stubs: {
        User,
        Loader,
        Error,
      },
    },
  });

  it('Отображение списка пользователей', async () => {
    jest.spyOn(store, 'dispatch').mockImplementation(() => new Promise((res) => {
      res(Promise.resolve(exampleUsers));
    }));

    const wrapper = getWrapper();

    await flushPromises();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Отображение во время загрузки', async () => {
    jest.spyOn(store, 'dispatch').mockImplementation((type): Promise<void> => new Promise((resolve) => {
      if (type === storeTypes.EActions.FETCH_USERS) {
        setTimeout(() => {
          resolve();
        }, 10000);
      } else {
        resolve();
      }
    }));

    const wrapper = getWrapper();

    await nextTick();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Отображение во время ошибки при загрузке пользователей', async () => {
    const wrapper = getWrapper();

    jest.spyOn(store, 'dispatch').mockImplementation(() => Promise.reject(new Error('Ошибка')));

    await flushPromises();

    expect(wrapper.html()).toMatchSnapshot();
  });
});
