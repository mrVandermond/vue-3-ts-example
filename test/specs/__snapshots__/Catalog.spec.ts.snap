// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Catalog component > Отображение во время загрузки 1`] = `"<div></div>"`;

exports[`Catalog component > Отображение во время ошибки при загрузке пользователей 1`] = `
"<div>
  <div class=\\"error\\"><img src=\\"assets/error.png\\" class=\\"error-image\\" alt=\\"error-image\\">
    <div class=\\"\\">
      <p class=\\"title-text title-text__bold\\"> Сервер не отвечает </p>
      <p class=\\"title-text\\"> Уже работаем над этим </p>
    </div>
  </div>
</div>"
`;

exports[`Catalog component > Отображение списка пользователей 1`] = `
"<div>
  <list-item-stub isinternal=\\"false\\" isloadingcontent=\\"false\\" isfailloadingcontent=\\"false\\" class=\\"user\\"></list-item-stub>
  <list-item-stub isinternal=\\"false\\" isloadingcontent=\\"false\\" isfailloadingcontent=\\"false\\" class=\\"user\\"></list-item-stub>
</div>"
`;
